[BE] Enable C++ stacktraces for MultiProcessTestCase

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69175

Shows C++ stacktraces for python distributed tests that inherit from
MultiProcessTestCase. Closes https://github.com/pytorch/pytorch/issues/69168

Example trace - 

```
File "/data/sandcastle/boxes/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_nccl_spawn#binary,link-tree/torch/nn/parallel/distributed.py", line 904, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: test error
Exception raised from rebuild_buckets at caffe2/torch/csrc/distributed/c10d/reducer.cpp:1652 (most recent call first):
# 0  c10::get_backtrace[abi:cxx11](unsigned long, unsigned long, bool)
# 1  std::_Function_handler<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (), c10::(anonymous namespace)::GetFetchStackTrace()::$_0>::_M_invoke(std::_Any_data const&)
# 2  c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
# 3  c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*)
# 4  c10d::Reducer::rebuild_buckets()
# 5  pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}, bool, c10d::Reducer*, pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}&&, bool (*)(c10d::Reducer*), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(pybind11::detail::function_call&)#1}::operator()(pybind11::detail::function_call) const
```

Differential Revision: [D32736872](https://our.internmc.facebook.com/intern/diff/D32736872/)

rohan-varma committed:
[BE] Enable C++ stacktraces for MultiProcessTestCase

Shows C++ stacktraces for python distributed tests that inherit from
MultiProcessTestCase. Closes https://github.com/pytorch/pytorch/issues/69168

Differential Revision: [D32736872](https://our.internmc.facebook.com/intern/diff/D32736872/)

[ghstack-poisoned]

H-Huang commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -569,6 +569,8 @@ def run_test(
             # Windows and MacOS do not support the signal handlers.
             torch._C._set_print_stack_traces_on_fatal_signal(True)
+        # Show full C++ stacktraces when a Python error originating from C++ is raised.
+        os.environ["TORCH_SHOW_CPP_STACKTRACES"] = "1"
```
Is this env var also getting passed to all subprocesses?

rohan-varma replied:
yep it should be passed to subprocesses, but this is actually already in the subprocess fn anyways

H-Huang reviewed (approved):
looks good! Just 1 question

rohan-varma committed:
Update on "[BE] Enable C++ stacktraces for MultiProcessTestCase"


Shows C++ stacktraces for python distributed tests that inherit from
MultiProcessTestCase. Closes https://github.com/pytorch/pytorch/issues/69168

Example trace - 

```
File "/data/sandcastle/boxes/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_nccl_spawn#binary,link-tree/torch/nn/parallel/distributed.py", line 904, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: test error
Exception raised from rebuild_buckets at caffe2/torch/csrc/distributed/c10d/reducer.cpp:1652 (most recent call first):
# 0  c10::get_backtrace[abi:cxx11](unsigned long, unsigned long, bool)
# 1  std::_Function_handler<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (), c10::(anonymous namespace)::GetFetchStackTrace()::$_0>::_M_invoke(std::_Any_data const&)
# 2  c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
# 3  c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*)
# 4  c10d::Reducer::rebuild_buckets()
# 5  pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}, bool, c10d::Reducer*, pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}&&, bool (*)(c10d::Reducer*), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(pybind11::detail::function_call&)#1}::operator()(pybind11::detail::function_call) const
```

Differential Revision: [D32736872](https://our.internmc.facebook.com/intern/diff/D32736872/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE] Enable C++ stacktraces for MultiProcessTestCase"


Shows C++ stacktraces for python distributed tests that inherit from
MultiProcessTestCase. Closes https://github.com/pytorch/pytorch/issues/69168

Example trace - 

```
File "/data/sandcastle/boxes/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_nccl_spawn#binary,link-tree/torch/nn/parallel/distributed.py", line 904, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: test error
Exception raised from rebuild_buckets at caffe2/torch/csrc/distributed/c10d/reducer.cpp:1652 (most recent call first):
# 0  c10::get_backtrace[abi:cxx11](unsigned long, unsigned long, bool)
# 1  std::_Function_handler<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (), c10::(anonymous namespace)::GetFetchStackTrace()::$_0>::_M_invoke(std::_Any_data const&)
# 2  c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
# 3  c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*)
# 4  c10d::Reducer::rebuild_buckets()
# 5  pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}, bool, c10d::Reducer*, pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(pybind11::cpp_function::initialize<bool, c10d::Reducer, , pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::call_guard<pybind11::gil_scoped_release> >(bool (c10d::Reducer::*)(), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(c10d::Reducer*)#1}&&, bool (*)(c10d::Reducer*), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, pybind11::call_guard<pybind11::gil_scoped_release> const&)::{lambda(pybind11::detail::function_call&)#1}::operator()(pybind11::detail::function_call) const
```

Differential Revision: [D32736872](https://our.internmc.facebook.com/intern/diff/D32736872/)

[ghstack-poisoned]
